pkg/controller/rolloutblock/rolloutblock_controller.go

```diff
@@ -0,0 +1,622 @@
+		if c.rolloutblockWorkqueue.NumRequeues(rolloutBlockUpdater) >= maxRetries {
+			// Drop this update out of the workqueue and thus reset its
+			// backoff. This limits the time a "broken" object can hog a worker.
+			glog.Warningf("Update %q for RolloutBlock has been retried too many times, dropping from the queue", rolloutBlockUpdater)
+			c.rolloutblockWorkqueue.Forget(rolloutBlockUpdater)
+
+			return true
+		}
+
+		c.rolloutblockWorkqueue.AddRateLimited(rolloutBlockUpdater)
+
+		return true
+	}
+
+	glog.V(4).Infof("Successfully synced RolloutBlock Updater %q", rolloutBlockUpdater)
+	c.rolloutblockWorkqueue.Forget(obj)
+
+	return true
+}
+
+func (c *Controller) syncRolloutBlock(rolloutBlockUpdater RolloutBlockUpdater) error {
+	key := rolloutBlockUpdater.RolloutBlockKey
+	updaterType := rolloutBlockUpdater.UpdaterType
+	if updaterType == NewRolloutBlockObject {
+		return c.syncNewRolloutBlockObject(key)
+	}
+
+	overridingKey := rolloutBlockUpdater.OverridingObjectKey
+	if len(key) == 0 || len(overridingKey) == 0 {
+		return nil
+	}
+
+	var err error = nil
+	isDeletedObject := rolloutBlockUpdater.IsDeletedObject
+
+	if isDeletedObject {
+		switch updaterType {
+		case OverridingRelease:
+			err = c.removeReleaseFromRolloutBlockStatus(overridingKey, key)
+		case OverridingApplication:
+			err = c.removeAppFromRolloutBlockStatus(overridingKey, key)
+		}
+	} else {
+		switch updaterType {
+		case OverridingRelease:
+			err = c.addReleaseToRolloutBlockStatus(overridingKey, key)
+		case OverridingApplication:
+			err = c.addApplicationToRolloutBlockStatus(overridingKey, key)
+		}
+	}
+
+	return err
+}
+
+func (c *Controller) syncNewRolloutBlockObject(key string) error {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	rolloutBlock, err := c.rolloutBlockLister.RolloutBlocks(ns).Get(name)
+	if err != nil {
+		return err
+	}
+
+	if ns == shipper.ShipperNamespace {
+		apps, err := c.applicationLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if err = c.addMultipleApplicationsToRolloutBlocks(apps, key, rolloutBlock); err != nil {
+			return err
+		}
+
+		rels, err := c.releaseLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if err = c.addMultipleReleasesToRolloutBlocks(rels, key, rolloutBlock); err != nil {
+			return err
+		}
+	} else {
+		apps, err := c.applicationLister.Applications(ns).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if err = c.addMultipleApplicationsToRolloutBlocks(apps, key, rolloutBlock); err != nil {
+			return err
+		}
+
+		rels, err := c.releaseLister.Releases(ns).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if err = c.addMultipleReleasesToRolloutBlocks(rels, key, rolloutBlock); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) addMultipleReleasesToRolloutBlocks(releases []*shipper.Release, rolloutBlockKey string, rolloutBlock *shipper.RolloutBlock) error {
```

Comment:
Consider changing the method name to:
```go
addReleasesToRolloutBlocks(blockKey, block, ...*shipper.Release)
```
In this case you can reuse this method for adding a singular app to the block.